02-intermediate-ruby/intro-to-minitest.md

```diff
@@ -0,0 +1,378 @@
+NameError: uninitialized constant Bill
+    bill_spec.rb:9:in `block (2 levels) in <main>'
+
+
+bin/rails test bill_spec.rb:8
+
+
+
+Finished in 0.003340s, 299.4012 runs/s, 0.0000 assertions/s.
+
+  1) Error:
+bill#test_0001_can be created:
+NameError: uninitialized constant Bill
+    bill_spec.rb:9:in `block (2 levels) in <main>'
+
+1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
+
+
+
+Finished in 0.457662s, 2.1850 runs/s, 0.0000 assertions/s.
+1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
+```
+
+**This is a good thing.**  We have our first **red** test.  There's an error because we haven't created the Bill class yet.
+
+####  Step 4:  Write Code To Make It Pass
+
+Lets make the test pass by creating a Bill class.
+
+```ruby
+# bill.rb
+
+class Bill
+
+end
+```
+
+Now `ruby bill_spec.rb` gets us:
+
+```bash
+Run options: --seed 32657
+
+# Running:
+
+Run options: --seed 32657
+
+# Running:
+
+..
+
+Finished in 0.001089s, 918.2736 runs/s, 918.2736 assertions/s.
+
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+Finished in 0.454822s, 2.1987 runs/s, 2.1987 assertions/s.
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+```
+
+Now we have our first **green/passing** test.
+
+There are a [number of expectations](http://mattsears.com/articles/2011/12/10/minitest-quick-reference/) in Minitest beyond the `must_equal` method.
+
+**Exercise** Look at the list of expectations above and revise the `bill.class.must_equal Bill` to use the `must_be_instance_of` expectation instead.
+
+
+#### A Word on Parentheses
+
+In the code above, we are calling `must_equal` without using parentheses `bill.class.must_equal Bill`.  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide) suggest that it's good coding style to put parentheses around method arguments **except** for methods which are part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.
+
+So you shouldn't put parentheses around the arguments to `must` method arguments, but you **should** around your own methods.
+
+### Adding Functionality
+
+We are going to require that Bill must be instantiated with a list (array) of hashes.  Each item will have a name and price.  So we will revise the test to look like this:
+
+```ruby
+# bill_test.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'bill'
+
+describe "bill" do
+  it "can be created" do
+    bill = Bill.new [{name: 'Coffee', price: 4.75}, {name: 'Pancakes', price: 8.75}]
+    # the class of @bill should be Die
+    bill.class.must_equal Bill
+  end
+end
+```
+
+We also need to ensure that if Bill.new is called without an argument it raises an error.  
+
+So we will add another `it` block and use the `must_raise` expectation.
+
+```ruby
+it "will raise an error if created with an invalid argument" do
+  proc {
+    bill = Bill.new
```

Comment:
Rather than "with an invalid argument", could you say something more descriptive? Perhaps "without a list of purchases".

Reply:
I would also argue that this isn't an interesting behavior to test, since throwing an `ArgumentError` when you forget a parameter is baked into Ruby. There's no line of code you could remove or change to make this test red.

A more interesting scenario might be "what happens when the provided list is empty?"